### PR TITLE
Clean up duplicated webhooks when istioctl fails

### PR DESCRIPTION
**Pull request by cnvergence**
- Clean up conflicted webhooks when istioctl fails to unblock reconciliations.

See also: 
- https://github.com/kyma-project/kyma/issues/17054

### CODE REVIEW

File: pkg/reconciler/instances/istio/webhooks/webhooks.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+func isDefaultRevisionDeactivated(ctx context.Context, client kubernetes.Interface) bool {
+	webhooks, err := tag.GetWebhooksWithRevision(ctx, client, tag.DefaultRevisionName)
+	if err != nil {
+		return true
+	}
+	if len(webhooks) == 0 {
+		return true
+	}
+
+	webhook := webhooks[0]
```
Comment (triffer) [inline pkg/reconciler/instances/istio/webhooks/webhooks.go:73]: Why can we rely on the order of the webhooks and always check the first webhook? As far as I understand this `GetWebhooksWithRevision` will return `istio-sidecar-injector` and `istio-revision-tag-default` and we only want to verify that all webhooks in the `istio-sidecar-injector` webhook configuration are deactivated. Should we add a filter for the name to make it more clear?

File: pkg/reconciler/instances/istio/webhooks/webhooks_test.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		// then
+		require.NoError(t, err)
+		got, err := client.AdmissionregistrationV1().MutatingWebhookConfigurations().Get(ctx, taggedWhName, metav1.GetOptions{})
+		require.NoError(t, err)
+		require.Equal(t, taggedWh, got)
+	})
+	t.Run("should delete conflicted tagged webhook if old one is not deactivated", func(t *testing.T) {
+		// given
+		defaultWh := createDefaultMutatingWebhookWithSelector(defaultWhName, "istio.io/rev", validSelector)
+		taggedWh := createDefaultMutatingWebhookWithSelector(taggedWhName, tag.IstioTagLabel, validSelector)
```
Comment (triffer) [inline pkg/reconciler/instances/istio/webhooks/webhooks_test.go:84]: We might want to update the test setup as both webhook configurations are created with the label "istio.io/rev" with the value "default".

File: pkg/reconciler/instances/istio/webhooks/webhooks.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	err = retry.Do(func() error {
+		webhooks, err := tag.GetWebhooksWithTag(ctx, kubeClient, tag.DefaultRevisionName)
+		if err != nil {
+			return err
+		}
+
+		if !isDefaultRevisionDeactivated(ctx, kubeClient) && len(webhooks) > 0 {
+			for _, wh := range webhooks {
```
Comment (triffer) [inline pkg/reconciler/instances/istio/webhooks/webhooks.go:46]: Maybe we can add a comment like this to make it more clear why we can delete this webhooks:
"As the default revision is not deactivated and handles the injection, we are safe to delete all other webhook configurations that were created during the failed installation."

File: pkg/reconciler/instances/istio/webhooks/webhooks_test.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		client := fake.NewSimpleClientset(defaultWh, taggedWh)
+		provider := &clientsetmocks.Provider{}
+		provider.On("RetrieveFrom", mock.AnythingOfType("string"), mock.AnythingOfType("*zap.SugaredLogger")).Return(client, nil)
+		// when
+		err := DeleteConflictedDefaultTag(ctx, provider, kubeConfig, log)
+
+		// then
+		require.NoError(t, err)
+		got, err := client.AdmissionregistrationV1().MutatingWebhookConfigurations().Get(ctx, taggedWhName, metav1.GetOptions{})
```
Comment (triffer) [inline pkg/reconciler/instances/istio/webhooks/webhooks_test.go:93]: Adding an assert that the default webhook is not deleted might be a good idea. What do you think?

File: pkg/reconciler/instances/istio/webhooks/webhooks.go (hunk 3 above)
Comment (cnvergence) [reply pkg/reconciler/instances/istio/webhooks/webhooks.go:46]: I will add the comment

File: pkg/reconciler/instances/istio/webhooks/webhooks_test.go (hunk 2 above)
Comment (cnvergence) [reply pkg/reconciler/instances/istio/webhooks/webhooks_test.go:84]: oh, I think I missed that information

File: pkg/reconciler/instances/istio/webhooks/webhooks.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+		return true
+	}
+	if len(webhooks) == 0 {
+		return true
+	}
+
+	webhook := webhooks[0]
+	for i := range webhook.Webhooks {
+		wh := webhook.Webhooks[i]
+		if fmt.Sprint(wh.NamespaceSelector.MatchLabels) == fmt.Sprint(deactivatedLabel) || fmt.Sprint(wh.ObjectSelector) == fmt.Sprint(deactivatedLabel) {
```
[the remark this answers is not in the document]
Comment (cnvergence) [reply pkg/reconciler/instances/istio/webhooks/webhooks.go:76]: tbh, any occurence of never-match label would be enough to count that MWC as deactivated
I can add AND with no problem at all.

File: pkg/reconciler/instances/istio/actions/performer.go [hunk 6]
```diff
@@ -197,6 +198,11 @@ func (c *DefaultIstioPerformer) Install(context context.Context, kubeConfig, ist
 
 	err = commander.Install(mergedCNI, kubeConfig, logger)
 	if err != nil {
+		// In case of error in the istioctl, old mutatingwebhook won't be deactivated, which will block later reconciliations.
+		err2 := webhooks.DeleteConflictedDefaultTag(context, c.provider, kubeConfig, logger)
+		if err2 != nil {
```
Comment (triffer) [inline pkg/reconciler/instances/istio/actions/performer.go:203]: If we don't want to wrap err2 into err I think we should at least log err2 on error level.

**Review by triffer — APPROVED** [review]

**Review by ruanxin — APPROVED** [review]

**Comment by Ressetkk** [comment]
As per https://github.com/kyma-project/kyma/issues/17030#issuecomment-1471915673:
> The issue has been fixed on main. Currently we will override the result of the job to be green. As the fix has been already backported to release branch 2.12. Check [here](https://github.com/kyma-project/kyma/commit/86447847e0919cb1062dfa1a98e6c3b8ca4d6bcb). Once we release 2.12.1 then the job should work fine and we should not need to override the result anymore.

/override pre-main-reconciler-cli-kyma-prev-to-last-release-upgrade-k3d

**Comment by Ressetkk** [comment]
/override rakesh-garimella_test_of_prowjob_pre-main-reconciler-cli-kyma-p

**Pull request merged**